### DISCUSSION TITLE
Accuracy of dual variable (Lagrange multiplier, a.k.a, costate) by IPOPT

### DISCUSSION THREAD

**Discussion by sixtosevenf** [casadi/casadi#3457] [Q&A]
Hello,

I am trying to investigate Lagrange multiplier (costate) of an optimal control problem (LQR problem). The LQR problem is defined with a cost function $J = \frac{1}{2}x^{\top}Qx + \frac{1}{2}u^{\top}Ru$ and a system dynamics equality constraint $\dot{x} = Ax+Bu$.

By solving the optimal control problem using CasADi with IPOPT, I obtained the optimal control trajectories $u_{CasADi}^* (t)$ and the Lagrange multiplier (costate) trajectories $\lambda^* (t)$ with the command `opti.lam_g`.

From the KKT condition of the LQR optimal control problem, the optimal control trajectories can be derived as $u_{KKT}^* (t) = -R^{-1}B^{\top}\lambda^* (t)$.

I compared the optimal control trajectories $u_{CasADi}^* (t)$ and $u_{KKT}^* (t) = -R^{-1}B^{\top}\lambda^* (t)$ with the $\lambda^*$ obtained from CasADi `opti.lam_g` and those trajectories are very different.

Am I doing something wrong? or the issue comes from the accuracy of $\lambda^{*}(t)$ of CasADi `opti.lam_g`?
If the latter is trouble, how can I improve the accuracy of Lagrange multiplier result?

Any comments would be appreciated.

**Comment by brunoml5** [comment]
Hi!

I'm not a huge expert, but I believe that the problem is not casadi, but the barrier method used in ipopt and similar solvers. Since barrier methods uses relaxations for all constraints, they are not very precise on the dual variables.

Take a look at this [article](https://www.sciencedirect.com/science/article/pii/S2405896318327137) from J. Andersson about sensitivity analysis in casadi. They talk about sensitivity analysis in casadi using an active-set method. The idea of the method is to solve the NLP using a robust method such ipopt and, using the solution as a starting point, solve the problem again with the active-set method.

Take a look at this [blog post](https://web.casadi.org/blog/nlp_sens/) for an example on how to implement the method.

**Reply by sixtosevenf** [reply]
Thank Bruno for your comment.

I have read some related posts and the paper. I tried the example in the [blog post](https://web.casadi.org/blog/nlp_sens/) and checked the sensitivity and the Lagrange multiplier, respectively.

Based on the principle (the constraint relaxation of IPOPT or active-set method), I think the solvers (IPOPT and qrqp) may be related to the Lagrange multiplier.

For now, I do not understand the reason but the calculated Lagrange multiplier by IPOPT and qrqp are almost same even though the sensitivity values are quite different. So, I am not sure how to interpret the situation.

**Comment by jaeandersson** [comment]
It's not clear to me if it's an issue of accuracy or not. Are the value correct within some tolerance or just completely wrong? I wonder if there is a confusion about the definition of `opti.lam_g`: My understanding is that it just maps to the NLP solver `lam_g`, i.e. the dual variables corresponding to nonlinear constraints of the NLP (@jgillis can correct me since he made the Opti interface). These can map more or less trivially to each other, but it may depend on the discretization scheme.

**Reply by sixtosevenf** [reply]
Thank Joel for your comment.

In terms of `lam_g`, my understanding is the same and it is the Lagrange multiplier I tried to investigate (@jgillis please correct me if my understanding is wrong).

The calculated Lagrange multiplier value is quite different but I found that there may be some scaling factors for Lagrange multiplier.

I derived the Lagrange multiplier from KKT, which are correct values and compared the values with the obtained Lagrange multiplier by IPOPT. See the below figures (not scaled vs. scaled)

![image](https://github.com/casadi/casadi/assets/148242362/77e91f16-f43b-4dfb-994c-205249029c89)
(Fig.1 Not scaled Lagrange multipliers)

![image](https://github.com/casadi/casadi/assets/148242362/6c5909f9-0835-469c-85b1-d19a4185b1e0)
(Fig. 2 Scaled Lagrange multipliers)

Could you (@jaeandersson and @jgillis) please confirm that there is scaling or other things related to the Lagrange multiplier computation?

**Reply by brunoml5** [reply]
Hi!

Certainly IPOPT uses scaling of the NLP. Try changing the [options](https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_NLP_Scaling), maybe change `nlp_scaling_method` to 'none' and see if it makes sense to you.

It is also confirmed in this Casadi [blog post](https://web.casadi.org/blog/nlp-scaling/) with this comment:

> [...] In IPOPT, objective and constraints (but not variables) are scaled automatically, using first-order sensitivities at the initial guess. This is vital for the above example to work. Without it, we need to scale objective and constraints ourselves.

I hope it helps!

**Reply by sixtosevenf** [reply]
Thanks, Bruno. I applied and tried changing the options but it did not work, unfortunately. It does not seem like `nlp_scaling_method `is related to the Lagrange multiplier computation.

**Comment by jaeandersson** [comment]
> Thank Joel for your comment.
> 
> In terms of `lam_g`, my understanding is the same and it is the Lagrange multiplier I tried to investigate (@jgillis please correct me if my understanding is wrong).
> 
> The calculated Lagrange multiplier value is quite different but I found that there may be some scaling factors for Lagrange multiplier.
> 
> I derived the Lagrange multiplier from KKT, which are correct values and compared the values with the obtained Lagrange multiplier by IPOPT. See the below figures (not scaled vs. scaled)
> 
> ![image](https://user-images.githubusercontent.com/148242362/282860458-77e91f16-f43b-4dfb-994c-205249029c89.png) (Fig.1 Not scaled Lagrange multipliers)
> 
> ![image](https://user-images.githubusercontent.com/148242362/282860790-6c5909f9-0835-469c-85b1-d19a4185b1e0.png) (Fig. 2 Scaled Lagrange multipliers)
> 
> Could you (@jaeandersson and @jgillis) please confirm that there is scaling or other things related to the Lagrange multiplier computation?

I don't think scaling factors within IPOPT matters, unless you are looking explicitly at the solver logs from IPOPT. To resolve it, I would plug in some values for A, B, C, D, Q and R (maybe start with dimension 1) and see if you still have the mismatch. Then compare the answer with the analytical solution. Also keep in mind that there is `lam_x`, corresponding to simple bounds (lbx and ubx). If the simple bounds become active at some point, you need to include `lam_x` in the calculation of the optimal control.

**Reply by sixtosevenf** [reply]
Thank Joel for your comment.

Actually, I found that the scaling is the discretization sampling time $\Delta t$. In the Lagrange multiplier computation, is it computed as a continuous variable? (e.g. $\dot{\lambda}(t) = \frac{\lambda_{k+1} - \lambda_{k}}{\Delta t}$).

I do not know the internal procedure but I can get the correct Lagrange multiplier with the sampling time $\Delta t$ and the accuracy of Lagrange multiplier decreases as the sampling time $\Delta t$ increases.

Could you give us any comment on this?

**Reply by brunoml5** [reply]
I had an idea here. 

What is your objective function? Do you use the sum of the stage costs or do you implement the integral?

I noticed that you have a continous system, so maybe you need to implement the stage cost as a integral. Use a quadrature function for that. You can see how to implement this in the example pack.

Also, maybe share your code to us so that we can take a look.

**Comment by jaeandersson** [comment]
> Thank Joel for your comment.
> 
> Actually, I found that the scaling is the discretization sampling time Δt. In the Lagrange multiplier computation, is it computed as a continuous variable? (e.g. λ˙(t)=λk+1−λkΔt).
> 
> I do not know the internal procedure but I can get the correct Lagrange multiplier with the sampling time Δt and the accuracy of Lagrange multiplier decreases as the sampling time Δt increases.
> 
> Could you give us any comment on this?

The NLP solvers in CasADi don't have any notion of time. So it must be in the mapping from discretized solution to continuous time solution. Which may be inside Opti if you're using the dynamic discretization from there
